From 35cbd2d32df76644b2259da7c31f5353766b858a Mon Sep 17 00:00:00 2001
From: Kirill Kapranov <kirill.kapranov@compulab.co.il>
Date: Wed, 16 Sep 2020 21:40:06 +0300
Subject: [PATCH 50/50] Enable dual boot Android and Linux

The boot loader allows to boot a Linux image from external storage; if not found
try to boot Android from a boot storage.

Signed-off-by: Kirill Kapranov <kirill.kapranov@compulab.co.il>
---
 configs/ucm-imx8m-mini-android_defconfig | 75 ++++++++++++++++++++++++++++++++
 include/configs/ucm-imx8m-mini.h         |  3 ++
 include/configs/ucm-imx8m-mini_android.h | 60 +++++++++++++++++++++++++
 3 files changed, 138 insertions(+)
 create mode 100644 configs/ucm-imx8m-mini-android_defconfig
 create mode 100644 include/configs/ucm-imx8m-mini_android.h

diff --git a/configs/ucm-imx8m-mini-android_defconfig b/configs/ucm-imx8m-mini-android_defconfig
new file mode 100644
index 0000000000..e39024764d
--- /dev/null
+++ b/configs/ucm-imx8m-mini-android_defconfig
@@ -0,0 +1,75 @@
+CONFIG_ARM=y
+CONFIG_ARCH_IMX8M=y
+CONFIG_SYS_TEXT_BASE=0x40200000
+CONFIG_SYS_MALLOC_F_LEN=0x2000
+CONFIG_TARGET_UCM_IMX8M_MINI=y
+CONFIG_SPL_MMC_SUPPORT=y
+CONFIG_IMX_OPTEE=y
+CONFIG_FLASH_MCUFIRMWARE_SUPPORT=y
+CONFIG_DEFAULT_DEVICE_TREE="ucm-imx8m-mini"
+CONFIG_FIT=y
+CONFIG_SPL_LOAD_FIT=y
+CONFIG_SYS_EXTRA_OPTIONS="IMX_CONFIG=arch/arm/mach-imx/spl_sd.cfg,SPL_TEXT_BASE=0x7E1000,ANDROID_SUPPORT"
+CONFIG_DEFAULT_FDT_FILE="ucm-imx8m-mini.dtb"
+CONFIG_ARCH_MISC_INIT=y
+CONFIG_SPL=y
+CONFIG_SPL_BOARD_INIT=y
+CONFIG_HUSH_PARSER=y
+CONFIG_FASTBOOT=y
+CONFIG_FSL_FASTBOOT=y
+CONFIG_FASTBOOT_BUF_ADDR=0x40480000
+CONFIG_FASTBOOT_BUF_SIZE=0x19000000
+CONFIG_FASTBOOT_FLASH=y
+CONFIG_FASTBOOT_FLASH_MMC_DEV=0
+CONFIG_CMD_EEPROM=y
+CONFIG_CMD_EEPROM_LAYOUT=y
+CONFIG_CMD_MEMINFO=y
+CONFIG_CMD_MEMTEST=y
+CONFIG_CMD_GPIO=y
+CONFIG_CMD_I2C=y
+CONFIG_CMD_CACHE=y
+# CONFIG_CMD_LED is not set
+CONFIG_CMD_REGULATOR=y
+CONFIG_CMD_EXT2=y
+CONFIG_CMD_EXT4=y
+CONFIG_CMD_FAT=y
+CONFIG_CMD_FS_GENERIC=y
+CONFIG_EFI_PARTITION=y
+CONFIG_PARTITION_TYPE_GUID=y
+CONFIG_OF_CONTROL=y
+CONFIG_IMX8M_LPDDR4=y
+CONFIG_ENV_IS_IN_MMC=y
+CONFIG_DM_GPIO=y
+CONFIG_SYS_I2C_MXC=y
+CONFIG_LED=y
+CONFIG_LED_STATUS=y
+CONFIG_LED_STATUS_GPIO=y
+CONFIG_LED_STATUS0=y
+CONFIG_LED_STATUS_BIT=12
+CONFIG_DM_MMC=y
+CONFIG_DM_SPI_FLASH=y
+CONFIG_SPI_FLASH=y
+CONFIG_SPI_FLASH_STMICRO=y
+CONFIG_DM_ETH=y
+CONFIG_USB_HOST_ETHER=y
+CONFIG_USB_ETHER_ASIX88179=y
+CONFIG_PINCTRL=y
+CONFIG_PINCTRL_IMX8M=y
+CONFIG_DM_REGULATOR=y
+CONFIG_DM_REGULATOR_FIXED=y
+CONFIG_DM_REGULATOR_GPIO=y
+CONFIG_DM_SPI=y
+CONFIG_DM_THERMAL=y
+CONFIG_NXP_TMU=y
+CONFIG_USB=y
+CONFIG_DM_USB=y
+CONFIG_USB_EHCI_HCD=y
+CONFIG_USB_GADGET=y
+CONFIG_USB_GADGET_MANUFACTURER="FSL"
+CONFIG_USB_GADGET_VENDOR_NUM=0x0525
+CONFIG_USB_GADGET_PRODUCT_NUM=0xa4a5
+CONFIG_SDP_LOADADDR=0x40400000
+CONFIG_FS_FAT=y
+CONFIG_LZ4=y
+CONFIG_SPL_GPIO_SUPPORT=y
+CONFIG_USB_GADGET_DOWNLOAD=y
diff --git a/include/configs/ucm-imx8m-mini.h b/include/configs/ucm-imx8m-mini.h
index cd1ed129f9..68edb32cb0 100644
--- a/include/configs/ucm-imx8m-mini.h
+++ b/include/configs/ucm-imx8m-mini.h
@@ -140,6 +140,8 @@
 	"setenv bootargs console=${console} root=/dev/mmcblk1p2;\0" \
 	"usb_ul=usb start; setenv iface usb; setenv dev 0; setenv part 1;" \
 	"setenv bootargs console=${console} root=/dev/sda2;\0" \
+	"ulandroid=if load mmc ${dev}:5 $loadaddr default.prop; then env def -a;" \
+	"boota; fi\0" \
 	"ulbootscript=load ${iface} ${dev}:${part} ${loadaddr} ${script};\0" \
 	"ulimage=load ${iface} ${dev}:${part} ${loadaddr} ${image}\0" \
 	"ulfdt=if test ${boot_fdt} = yes || test ${boot_fdt} = try; then " \
@@ -183,6 +185,7 @@
 			"fi; " \
 		"fi; " \
 	"done; " \
+	"run ulandroid; "\
 	"usb start; ums 0 mmc ${mmcdev};"
 
 /* Link Definitions */
diff --git a/include/configs/ucm-imx8m-mini_android.h b/include/configs/ucm-imx8m-mini_android.h
new file mode 100644
index 0000000000..1191bec3b8
--- /dev/null
+++ b/include/configs/ucm-imx8m-mini_android.h
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2018 NXP
+ *
+ * SPDX-License-Identifier:	GPL-2.0+
+ */
+
+#ifndef __UCM_IMX8M_MINI_ANDROID_H
+#define __UCM_IMX8M_MINI_ANDROID_H
+
+#define CONFIG_BCB_SUPPORT
+#define CONFIG_CMD_READ
+
+#define CONFIG_ANDROID_AB_SUPPORT
+#define CONFIG_AVB_SUPPORT
+#define CONFIG_SUPPORT_EMMC_RPMB
+#define CONFIG_SYSTEM_RAMDISK_SUPPORT
+#define CONFIG_AVB_FUSE_BANK_SIZEW 0
+#define CONFIG_AVB_FUSE_BANK_START 0
+#define CONFIG_AVB_FUSE_BANK_END 0
+#define CONFIG_FASTBOOT_LOCK
+#define FSL_FASTBOOT_FB_DEV "mmc"
+
+#ifdef CONFIG_SYS_MALLOC_LEN
+#undef CONFIG_SYS_MALLOC_LEN
+#define CONFIG_SYS_MALLOC_LEN           (96 * SZ_1M)
+#endif
+
+#define CONFIG_ANDROID_RECOVERY
+
+#define CONFIG_CMD_BOOTA
+#define CONFIG_SUPPORT_RAW_INITRD
+#define CONFIG_SERIAL_TAG
+
+/* Enable mcu firmware flash */
+#ifdef CONFIG_FLASH_MCUFIRMWARE_SUPPORT
+#define ANDROID_MCU_FRIMWARE_DEV_TYPE DEV_MMC
+#define ANDROID_MCU_FIRMWARE_START 0x500000
+#define ANDROID_MCU_FIRMWARE_SIZE  0x40000
+#define ANDROID_MCU_FIRMWARE_HEADER_STACK 0x20020000
+#endif
+
+#ifdef CONFIG_FSL_CAAM_KB
+#undef CONFIG_FSL_CAAM_KB
+#endif
+#define AVB_AB_I_UNDERSTAND_LIBAVB_AB_IS_DEPRECATED
+
+#ifdef CONFIG_IMX_TRUSTY_OS
+#define AVB_RPMB
+#define KEYSLOT_HWPARTITION_ID 2
+#define KEYSLOT_BLKS             0x1FFF
+#define NS_ARCH_ARM64 1
+
+#ifdef CONFIG_SPL_BUILD
+#undef CONFIG_BLK
+#endif
+
+#endif
+
+#endif /* __UCM_IMX8M_MINI_ANDROID_H */
+
-- 
2.11.0

